include/swift/ABI/Task.h

```diff
@@ -190,12 +210,362 @@ class AsyncTask : public HeapObject, public Job {
+          std::string str;
+          str.append("GroupStatus{ ");
+          str.append("R:");
+          str.append(std::to_string(readyTasks()));
+          str.append(" P:");
+          str.append(std::to_string(pendingTasks()));
+          str.append(" W:");
+          str.append(std::to_string(waitingTasks()));
+          str.append(" " + std::bitset<64>(status).to_string());
+          str.append(" }");
+          return str;
+        }
+
+        /// Initially there are no waiting and no pending tasks.
+        static const GroupStatus initial() {
+          return GroupStatus { 0 };
+        };
+    };
+
+    template<typename T>
+    class NaiveQueue {
+        std::queue<T> queue;
+
+    public:
+        NaiveQueue() = default;
+        NaiveQueue(const NaiveQueue<T> &) = delete ;
+        NaiveQueue& operator=(const NaiveQueue<T> &) = delete ;
+
+        NaiveQueue(NaiveQueue<T>&& other) {
+          queue = std::move(other.queue);
+        }
+
+        virtual ~NaiveQueue() { }
+
+        bool dequeue(T &output) {
+          if (queue.empty()) {
+            return false;
+          }
+          output = queue.front();
+          queue.pop();
+          return true;
+        }
+
+        void enqueue(const T item) {
+          queue.push(item);
+        }
+
+    };
+
+  private:
+
+    // TODO: move to lockless via the status atomic
+    mutable std::mutex mutex;
+
+    /// Used for queue management, counting number of waiting and ready tasks
+    std::atomic<unsigned long> status;
+
+    /// Queue containing completed tasks offered into this channel.
+    ///
+    /// The low bits contain the status, the rest of the pointer is the
+    /// AsyncTask.
+    NaiveQueue<ReadyQueueItem> readyQueue;
+//     mpsc_queue_t<ReadyQueueItem> readyQueue; // TODO: can we get away with an MPSC queue here once actor executors land?
+
+// NOTE: this style of "queue" is not very nice for the group,
+//        because it acts more like a stack, and we really want completion order
+//        for the task group, thus not using this style (which the wait queue does)
+//    std::atomic<ReadyQueueItem> readyQueue;
+
+    /// Queue containing all of the tasks that are waiting in `get()`.
+    ///
+    /// A task group is also a future, and awaits on the group's result *itself*
+    /// are enqueued on its future fragment.
+    ///
+    /// The low bits contain the status, the rest of the pointer is the
+    /// AsyncTask.
+    std::atomic<WaitQueueItem> waitQueue;
+
+    friend class AsyncTask;
+
+  public:
+    explicit GroupFragment()
+        : status(GroupStatus::initial().status),
+          readyQueue(),
+//          readyQueue(ReadyQueueItem::get(ReadyStatus::Empty, nullptr)),
+          waitQueue(WaitQueueItem::get(WaitStatus::Waiting, nullptr)) {}
+
+    /// Destroy the storage associated with the channel.
+    void destroy();
+
+    bool isEmpty() {
+      auto oldStatus = GroupStatus { status.load(std::memory_order_relaxed) };
```

Comment:
This may be my lack of C++ experience showing...

Or could I model this as the atomic being of type GroupStatus?